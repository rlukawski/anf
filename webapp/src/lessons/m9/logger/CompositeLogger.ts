import { LogLevel } from "api/logs";
import { Logger, LogParams } from "./Logger.interface";

export class CompositeLogger implements Logger {
  private loggers: Logger[] = [];

  add(logger: Logger) {
    this.loggers.push(logger);
    return this;
  }

  send(level: LogLevel, message: string, params?: LogParams) {
    for (const logger of this.loggers) {
      logger.send(level, message, params);
    }
  }

  debug(message: string, params?: LogParams) {
    for (const logger of this.loggers) {
      logger.debug(message, params);
    }
  }

  info(message: string, params?: LogParams) {
    for (const logger of this.loggers) {
      logger.info(message, params);
    }
  }

  warn(message: string, params?: LogParams) {
    for (const logger of this.loggers) {
      logger.warn(message, params);
    }
  }

  error(message: string, params?: LogParams) {
    for (const logger of this.loggers) {
      logger.error(message, params);
    }
  }
}
