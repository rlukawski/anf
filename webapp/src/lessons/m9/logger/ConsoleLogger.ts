/* eslint-disable no-console */
import { LogLevel } from "api/logs";
import { Logger, LogParams } from "./Logger.interface";

export class ConsoleLogger implements Logger {

  static methodMap = {
    DEBUG: console.debug,
    INFO: console.info,
    WARN: console.warn,
    ERROR: console.error,
  };

  send(level: LogLevel, message: string, params?: LogParams) {
    ConsoleLogger.methodMap[level](`${level}: ${message}`, params || "");
  }

  debug(message: string, params?: LogParams) {
    console.debug(`DEBUG: ${message}`, params || "");
  }

  info(message: string, params?: LogParams) {
    console.info(`INFO: ${message}`, params || "");
  }

  warn(message: string, params?: LogParams) {
    console.warn(`WARN: ${message}`, params || "");
  }

  error(message: string, params?: LogParams) {
    console.error(`ERROR: ${message}`, params || "");
  }
}
