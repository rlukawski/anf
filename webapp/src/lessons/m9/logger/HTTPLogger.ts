import { sendLog, LogLevel } from "api/logs";
import { Logger, LogParams } from "./Logger.interface";

export class HTTPLogger implements Logger {
  private sendLog = async (logLevel: LogLevel, message: string, params?: LogParams) => {
    return sendLog(logLevel, message, params ?? null);
  };

  send = (level: LogLevel, message: string, params?: LogParams) => {
    return this.sendLog(level, message, params);
  };

  debug = (message: string, params?: LogParams) => {
    return this.send("DEBUG", message, params);
  };

  info = (message: string, params?: LogParams) => {
    return this.send("INFO", message, params);
  };

  warn = (message: string, params?: LogParams) => {
    return this.send("WARN", message, params);
  };

  error = (message: string, params?: LogParams) => {
    return this.send("ERROR", message, params);
  };
}
