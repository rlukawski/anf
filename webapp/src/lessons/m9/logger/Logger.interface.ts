import { LogLevel } from "api/logs";

export type LogParams = {
  [index: string]: any;
}

export interface Logger {
  send(level: LogLevel, message: string, params?: LogParams): void;
  debug(message: string, params?: LogParams): void;
  info(message: string, params?: LogParams): void;
  warn(message: string, params?: LogParams): void;
  error(message: string, params?: LogParams): void;
}
