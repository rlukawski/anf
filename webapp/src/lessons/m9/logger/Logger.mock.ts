import { Logger, LogParams } from "./Logger.interface";

export class MockLogger implements Logger {

  send = jest.fn()
  name = "MockLogger"

  debug(message: string, params?: LogParams){
    this.send('DEBUG', message, params)
  }
  info(message: string, params?: LogParams){
    this.send('INFO', message, params)
  }
  warn(message: string, params?: LogParams){
    this.send('WARN', message, params)
  }
  error(message: string, params?: LogParams){
    this.send('ERROR', message, params)
  }

  get expectLogs(){
    return expect(this.send.mock.calls)
  }
};
