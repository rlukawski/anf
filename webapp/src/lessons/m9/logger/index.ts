export * from './Logger.interface'
export * from './LoggerProvider'
export * from './HTTPLogger'
