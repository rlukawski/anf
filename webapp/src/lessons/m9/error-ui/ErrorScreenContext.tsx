import React, { createContext, useCallback, useContext, useState } from 'react';
import { useLogger, useLogMessage } from '../logger';

import { ErrorModal } from './ErrorModal';
import { ErrorPage } from './ErrorPage';
import { UserErrorMessage } from './messages'

type ErrorScreenContextValue = {
  showError(screen: ErrorScreenState): void
  clear(): void
}

export const ErrorScreenContext = createContext<ErrorScreenContextValue | undefined>(undefined)

interface ErrorScreenProviderProps {}

type ErrorScreenState =
  | {
    layout: "NO_ERROR"
  }
  | {
    layout: "ERROR_PAGE"
    message: UserErrorMessage
  }
  | {
    layout: "ERROR_MODAL"
    message: UserErrorMessage
  }

export const ErrorScreenProvider = (props: React.PropsWithChildren<ErrorScreenProviderProps>) => {
  const [errorScreen, setErrorScreen] = useState<ErrorScreenState>({ layout: "NO_ERROR" })
  const { logger } = useLogger()

  const showError = useCallback((screen: ErrorScreenState) => {
    setErrorScreen(screen)
    if (screen.layout !== "NO_ERROR"){
      logger.error(`Error Page displayed`, screen.message);
    }
  }, [logger])
  const clear = useCallback(() => {
    setErrorScreen({ layout: "NO_ERROR" })
  }, [])

  let content: React.ReactNode = props.children;
  if (errorScreen.layout === "ERROR_MODAL") {
    content = (
      <>
        {props.children}
        <ErrorModal
          isOpen={true}
          errorMessage={errorScreen.message}
          onClose={clear}
        />
      </>
    );
  } else if (errorScreen.layout === "ERROR_PAGE") {
    content = <ErrorPage errorMessage={errorScreen.message} />;
  }

  return <ErrorScreenContext.Provider value={{ showError, clear }}>
    {content}
  </ErrorScreenContext.Provider>
}

export const useErrorScreen = () => {
  const ctx = useContext(ErrorScreenContext)
  if (!ctx){
    throw new Error("Component beyond ErrorScreenContext")
  }
  return ctx
}
