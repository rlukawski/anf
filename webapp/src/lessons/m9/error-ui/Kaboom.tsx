import React, { useEffect } from "react";

import { useErrorScreen } from "./ErrorScreenContext";
import { getUserErrorMessage } from "./messages";

interface KaboomSyncProps {
  shouldFail?: () => boolean;
  notify?(message: string): void
};

const throwAlways = () => true

export const KaboomSync = (props: React.PropsWithChildren<KaboomSyncProps>) => {
  const { shouldFail = throwAlways, notify, children } = props;
  if (shouldFail()) {
    notify?.("KaboomSync: fail");
    throw new Error("Kaboom!");
  }
  notify?.("KaboomSync: pass");
  return <>{children}</>;
};

interface KaboomAsyncProps {
  kaboomFn: () => Promise<unknown>
  layout: "ERROR_PAGE" | "ERROR_MODAL"
}

export const KaboomAsync = (props: KaboomAsyncProps) => {
  const { kaboomFn, layout } = props;

  const { showError } = useErrorScreen();
  useEffect(() => {
    kaboomFn().catch((error: unknown) => {
      showError({
        layout,
        message: getUserErrorMessage(error),
      });
    });
  }, [layout, kaboomFn, showError]);

  return <div>this will blink</div>;
};
