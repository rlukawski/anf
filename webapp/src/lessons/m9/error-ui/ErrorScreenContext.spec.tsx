import React from "react";
import { render } from "@testing-library/react";
import axios from "axios";
import { rest } from "msw";
import { setupServer } from "msw/node";

import { ErrorScreenProvider } from "./ErrorScreenContext";
import { LoggerProvider } from "../logger";
import { MockLogger } from "../logger/Logger.mock";
import { KaboomAsync } from "./Kaboom";

const handlers = [
  rest.get("/mocks/errors/network", async (req, res, ctx) => {
    return res.networkError("Failed to connect");
  }),
  rest.get(`/mocks/errors/410`, async (req, res, ctx) => {
    return res(ctx.status(410, `Backend error message 410`));
  }),
];
const server = setupServer(...handlers);

describe("ErrorScreen", () => {
  beforeAll(() =>
    server.listen({
      onUnhandledRequest: "warn",
    }),
  );
  afterEach(() => server.resetHandlers());
  afterAll(() => server.close());

  type TestCase = {
    name: string;
    kaboomFn: () => Promise<unknown>;
    expectedText: string;
  };
  const testcases: TestCase[] = [
    {
      name: "unknown error (TypeError)",
      kaboomFn: async () => {
        throw new TypeError("undefined is not a function");
      },
      expectedText: "Unexpected error occurred. Please contact Support.",
    },
    {
      name: "Network Error",
      kaboomFn: async () => axios.get("/mocks/errors/network"),
      expectedText: "Internet connection error. Check your wires!",
    },
    {
      name: "TIMEOUT",
      kaboomFn: async () => axios.get("/mocks/errors/410"),
      expectedText: "This resource is no longer available.",
    },
  ];

  for (const { name, kaboomFn, expectedText } of testcases) {
    it(`should display message "${expectedText}" when ${name} was thrown`, async () => {
      const { findByText } = render(
        <ErrorScreenProvider>
          <KaboomAsync kaboomFn={kaboomFn} layout="ERROR_PAGE" />
        </ErrorScreenProvider>
      );

      await findByText(expectedText);
    });
  }

  for (const { name, kaboomFn, expectedText } of testcases) {
    it(`should send ERROR log (${expectedText}) if ErrorPage was displayed`, async () => {
      const mockLogger = new MockLogger();
      const { findByText } = render(
        <LoggerProvider logger={mockLogger}>
          <ErrorScreenProvider>
            <KaboomAsync kaboomFn={kaboomFn} layout="ERROR_PAGE" />
          </ErrorScreenProvider>
        </LoggerProvider>
      );

      await findByText("Oops...");

      mockLogger.expectLogs.toMatchSnapshot();
    });
  }
});
