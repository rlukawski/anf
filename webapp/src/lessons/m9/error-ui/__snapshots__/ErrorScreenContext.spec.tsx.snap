// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorScreen should send ERROR log (Internet connection error. Check your wires!) if ErrorPage was displayed 1`] = `
Array [
  Array [
    "ERROR",
    "Error Page displayed",
    Object {
      "type": "CONNECTION_FAILURE",
    },
  ],
]
`;

exports[`ErrorScreen should send ERROR log (This resource is no longer available.) if ErrorPage was displayed 1`] = `
Array [
  Array [
    "ERROR",
    "Error Page displayed",
    Object {
      "type": "TIMEOUT",
    },
  ],
]
`;

exports[`ErrorScreen should send ERROR log (Unexpected error occurred. Please contact Support.) if ErrorPage was displayed 1`] = `
Array [
  Array [
    "ERROR",
    "Error Page displayed",
    Object {
      "type": "UNEXPECTED_ERROR",
    },
  ],
]
`;
